### PR TITLE
Make `runRandomSource` usable

### PR DESCRIPTION
**Pull request by infinisil**
Previously, using `runRandomSource`'s type required [`RandomSource`](https://hackage.haskell.org/package/random-source-0.3.0.6/docs/Data-Random-Source.html#t:RandomSource) `(Sem r)`, which seems impossible or very hard to implement to me. By changing it to require `Member (Lift m) r` along with `RandomSource m s`, it becomes easily usable, because the Monad `m` the `RandomSource` requires can be lifted to the `Lift` effect.

If there is an easy way to use `runRandomSource` without this change, please let me know.

### CODE REVIEW

File: polysemy-RandomFu/polysemy-RandomFu.cabal [hunk 1]
```diff
@@ -63,11 +63,13 @@ test-suite polysemy-RandomFu-test
   build-depends:
       base >=4.7 && <5
     , hspec >=2.0
+    , mwc-random
```
Comment (adamConnerSax) [inline polysemy-RandomFu/polysemy-RandomFu.cabal]: Please put bounds in here, even though it's only for the test.

File: polysemy-RandomFu/test/RandomFuSpec.hs [hunk 2]
```diff
@@ -14,6 +14,10 @@ import           Control.Monad.IO.Class         ( liftIO )
 import qualified Data.Random                   as R
 import qualified Data.Random.Source.PureMT     as R
 
+import           System.Random.MWC
```
Comment (adamConnerSax) [inline polysemy-RandomFu/test/RandomFuSpec.hs]: Please qualify these imports or import only the functions you use.  I made an exception for the polysemy bits themselves because that's consistent with how everything in polysemy is.

File: polysemy-RandomFu/src/Polysemy/RandomFu.hs [hunk 3]
```diff
@@ -89,11 +91,11 @@ runRandomIO = interpret $ \case
 ------------------------------------------------------------------------------
 -- | Run in 'IO', using the given 'R.PureMT' source, stored in an 'IORef'
 runRandomIOPureMT
-  :: MonadIO (Sem r)
+  :: Member (Lift IO) r
```
Comment (adamConnerSax) [inline polysemy-RandomFu/src/Polysemy/RandomFu.hs:94]: Why these changes?  I think the two versions are equivalent?

File: polysemy-RandomFu/src/Polysemy/RandomFu.hs [hunk 4]
```diff
@@ -89,11 +91,11 @@ runRandomIO = interpret $ \case
...
   => R.PureMT
   -> Sem (RandomFu ': r) a
   -> Sem r a
 runRandomIOPureMT source re =
-  liftIO (newIORef source) >>= flip runRandomSource re
+  sendM (newIORef source) >>= flip runRandomSource re
```
Comment (adamConnerSax) [inline polysemy-RandomFu/src/Polysemy/RandomFu.hs:99]: See above.  I think the ```MonadIO``` instance for ```Member (Lift IO) r => Sem r``` is just ```liftIO = sendM``` so these versions are equivalent.  Am I missing something?

**Review by adamConnerSax — CHANGES_REQUESTED** [review]
Thanks for this!  I hadn't actually tried to make a ```runRandomSource``` example and so I hadn't thought it through.  And I had wanted to make sure this worked with MWC so I appreciate the work and the PR on both counts!
Why the changes in the ```runRandomIOPureMT```?  I think your version is, effectively, the same as mine.
Other than that, I had just a couple of little things, PVP bounds and qualified imports.  But otherwise I'm happy to merge and release a new version.

File: polysemy-RandomFu/polysemy-RandomFu.cabal (hunk 1 above)
Comment (adamConnerSax) [reply polysemy-RandomFu/polysemy-RandomFu.cabal]: Sorry!  I mean in the package.yaml.

File: polysemy-RandomFu/src/Polysemy/RandomFu.hs (hunk 3 above)
Comment (infinisil) [reply polysemy-RandomFu/src/Polysemy/RandomFu.hs:94]: Because `runRandomIOPureMT` uses `runRandomSource`, which requires this constraint. Can't compile without this change.

File: polysemy-RandomFu/src/Polysemy/RandomFu.hs (hunk 4 above)
Comment (infinisil) [reply polysemy-RandomFu/src/Polysemy/RandomFu.hs:99]: Yeah this change isn't necessarily needed, but I thought it made more sense this way with the changed constraint above, since the `Lift` effect is usually used with `sendM`, while `liftIO` is part of `MonadIO`.

File: polysemy-RandomFu/polysemy-RandomFu.cabal (hunk 1 above)
Comment (infinisil) [reply polysemy-RandomFu/polysemy-RandomFu.cabal]: Turns out `mwc-random` isn't even needed.

Should I still put a bound for `vector` though? I'd probably use `vector >= 0.7` since that's the bound used by `mwc-random`.

File: polysemy-RandomFu/polysemy-RandomFu.cabal [hunk 5]
```diff
@@ -70,4 +70,5 @@ test-suite polysemy-RandomFu-test
     , random-fu >=0.2.5.0 && <0.3.0.0
     , random-source >=0.3 && <0.4.0.0
     , text >=1.1.0.0 && <1.3.0.0
+    , vector
```
Comment (adamConnerSax) [inline polysemy-RandomFu/polysemy-RandomFu.cabal]: Yeah, put bounds here as well.  And an upper bound too.  I've been doing that since it's good form though I think I will regret the maintenance headache it will bring...

File: polysemy-RandomFu/src/Polysemy/RandomFu.hs (hunk 3 above)
Comment (adamConnerSax) [reply polysemy-RandomFu/src/Polysemy/RandomFu.hs:94]: Right!  I hadn't looked carefully enough.

File: polysemy-RandomFu/src/Polysemy/RandomFu.hs (hunk 4 above)
Comment (adamConnerSax) [reply polysemy-RandomFu/src/Polysemy/RandomFu.hs:99]: This makes sense.  Thanks!

**Comment by infinisil** [comment]
Did that :)
